### ISSUE TITLE
Implement the child trie removal.

### ISSUE THREAD

**Issue by pepyakin** [paritytech/substrate#1865]
At the present, we have functions to manipulate and query child storage tries (e.g. [ext_child_storage_root](https://github.com/paritytech/substrate/blob/fdb3a8466fb14d4790b7e10fd79fceeaef00ef44/core/executor/src/wasm_executor.rs#L372)).

However, we would like to migrate contract storage to use tries instead of the main storage (https://github.com/paritytech/substrate/issues/1864). The lack of function for removing the entire trie is preventing us from implementation of this change.

**Comment by rphmeier** [comment]
@pepyakin I am assuming right now that all contracts share the same storage trie but have their items' keys prefixed by their address or some similar scheme.

If that's the case, one approach to do this would be by extending the `patricia_trie` implementation to allow removal of all nodes with a prefix. In principle we should just be able to chop off some branch and recurse through its children to delete from DB and this can all be done in the `patricia_trie` level efficiently.

Then the contracts module could just do `storage.remove_prefix(contract_address)` when deleting the contract.

**Comment by xlc** [comment]
Please make this works with [StorageDoubleMap ](https://github.com/paritytech/substrate/blob/master/srml/support/src/double_map.rs)

**Comment by sorpaas** [comment]
@pepyakin We currently have `ext_kill_child_storage` (https://github.com/paritytech/substrate/blob/master/core/executor/src/wasm_executor.rs#L224) that does this. However, it will remove all values sync and we may want to improve that to be an async GC.

**Comment by rphmeier** [comment]
@xlc I am talking about that at the trie level, not at the substrate storage level.

**Comment by cheme** [comment]
@rphmeier @pepyakin If I understand correctly choice has to be made to either go the old eth way (putting prefix account key in trie db key construction) or use a child trie.
TLDR;I do not see a huge difference between both, prefix could be more straightforward as it directly include the child trie in the trie (proof of an account db value in the trie is more straight forward but it is only a mater of adding an operation). Some optimization allowed by prefix are the same if using child trie.

In fact I believe using child trie has the advantages to make optimization more obvious, for instance when restoring an account set of data it can create its own trie and attach its second node (root is the extension of the prefix and second node is the branch) in the existing trie. Similarily for data access where you only need to start at prefix using the triedb from prefixed branch root. 
When using child trie you first have to get its root from main trie: here it is obviously more costy as you go down a full key but it is not fair to compare both (different address space, if I want similar perf I just store the child trie at first 96 bit address). Then you have to run your operation from this trie root and directly benefit from the previous optimization.
Similarly using child trie make possible db optimization more obvious: for rent case removal of all child trie you could end up simply truncating a db collection/column. But to be fair you can do the same with checking key prefix (kvdb need to be aware that the prefix is a functionally bound set of data). 

On restoring contract state, having a child trie makes thing super simple to : just inline build it with efficient iterative tree algo. But similarily you can do it with prefix: just build locally and attach in main trie at prefix location.

I also think that if the child trie db could maintains its size in the main trie (store root + size in main trie), we can also have extension usage for small number and child trie for bigger trie.
TLDR; I feel like child trie db makes things more direct and could be nicer (especially when thinking at the db level and for some optimization), but to be fair when considering a account prefix as a child trie (no collision) it can be use the same way (optimization for deletion and also batch account insert).

So at the end I mostly think that it should be related to the way we want to use the db (aka would using db collection/column for accounts be a thing (current substrate protos do include both 'storage_key' (unused with current implementation) and 'storage root' which may indicate such will).

**Comment by cheme** [comment]
@sorpaas async gc for such deletion seems like a good idea to me (if i understand correctly maintain a set of to be deleted child trie or prefix).

**Comment by pepyakin** [comment]
Ahhh, `ext_kill_child_storage` is what I had in mind exactly! Just overlooked it 🤦‍♂️For the start, sync deletion will do.

**Comment by cheme** [comment]
you go also got the `ext_clear_prefix` if going with prefix :)

**Comment by pepyakin** [comment]
Yeah, as @xlc mentioned we already use `ext_clear_prefix` in `StorageDoubleMap`. The idea is that for the storage rent mechanism (see #1864) we will need to be able to get the storage root of a contract's storage efficiently, and it seems to me that using a child trie is more efficient rather than using prefixed approach. Right?

**Comment by cheme** [comment]
with assumption that common prefix only contains contract kv (prefix long enough to avoid collision), you can see the tree build from prefixed branch is a subtree and probably use the same optimization (on one side you update the child tree root in the main tree, on the other side you attach a branch on the main tree). 
But you may also skip such optimization, and the best way to ensure things run at a contract level (in my opinion) would be to use child trie (plus I think it open some db possibilities). It would be interesting to implement rent on eth with such prefix based optimization.
Also it is true that some comment in StorageDoubleMap points to a design with a prefix for each account/contract (see choice of hashing algorithm).

**Comment by cheme** [comment]
Edit: this is not doable in eth context (prefix is not applied in the trie key, but only on the underlying db).
